![Les billets de Semaine52 / AuFilDuBoamp / DataMining des données de marchés publics au Bulletin officiel](https://raw.githubusercontent.com/Semaine52/AuFilDuBoamp/master/IMG/BILLETS_S52_AUFD_BANDEAU.jpg "Logo")

##### **Data Mining des données de la commande publique parues au Bulletin officiel des annonces des marchés publics (BOAMP 2015-2019)**
**Source open data : https://www.data.gouv.fr/fr/datasets/api-boamp-beta/**
# **Aujourd'hui : la Ville Fantôme de Cedex et autres Joyeusetés**
### *ou «Les ravages de l'absence de référentiel. [...]» (cit. @JoelGombin, Twitter, 19/09/2019)*

____

# **A - Constatations**

<strong>Vous vous êtes toujours demandé ce qui s'est passé à Toulouse en 2018 niveau marchés publics (si,si, jouez le jeu).</strong> 

<strong>Ainsi, vous récupérez toute la colonne «ville» extraite des annonces de l'année</strong>.
En fait, je vous l'offre...
    
<strong>Vous obtenez le résultat sous le pâté de code suivant:</strong>.

In [9]:
# IMPORTS DES OUTILS
from urllib.request import urlretrieve
import pandas as pd

# RÉGLAGE IPYTHON INTERACTVESHELL
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all" 

# RÉGLAGES PANDAS
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_colwidth', -1)
pd.set_option('display.max_rows', 2000)
pd.set_option('display.max_columns', 500)

# IMPORT ET LECTURE DES DONNÉES
URL = 'https://www.semaine52.com/AuFilDuBoamp/PKL/df_d01_08_ville_2018_no_index.pkl'
urlretrieve(URL, 'df_d01_08_ville_2018_no_index.pkl');
df = pd.read_pickle('df_d01_08_ville_2018_no_index.pkl')
a, _ = df.shape

print('*' * 100)
print(f"Nombre d'annonce au BOAMP en 2018: {a} ")
print(f"Ci-dessous, les cinq premières lignes du tableau de résultats:")

df.head()

('df_d01_08_ville_2018_no_index.pkl', <http.client.HTTPMessage at 0x57d5400>)

****************************************************************************************************
Nombre d'annonce au BOAMP en 2018: 155935 
Ci-dessous, les cinq premières lignes du tableau de résultats:


d01_08_ville
0  Magny-les-Hameaux    
1  Conflans Ste-Honorine
2  ANNECY               
3  Toulouse cedex 1     
4  Caudry

____

<strong>Toujours en confiance, à partir des 115 935 entrées de départ, vous réduisez votre recherche à Toulouse:</strong>

In [10]:
df_toulouse = df.loc[lambda df: df.d01_08_ville == 'Toulouse', :].reset_index(drop=True)
df_toulouse.shape
df_toulouse.head()

(795, 1)

d01_08_ville
0  Toulouse   
1  Toulouse   
2  Toulouse   
3  Toulouse   
4  Toulouse

**Vous auriez donc 795 annonces sur l'année 2018 à Toulouse en stock.**  
**Mais il y a un problème, on le voit plus haut, c'est le fameux «cedex».**
**On a donc perdu de l'information, et Toulouse...**

#### **Je vous épargne ici les laborieuses étapes de nettoyage, et élargis l'angle de vue pour vous montrer ceci:**



In [11]:
df_boamp_2018_classe_par_nom_de_ville = df.sort_values(by='d01_08_ville').reset_index(drop=True)
df_boamp_2018_classe_par_nom_de_ville.iloc[0:10]

d01_08_ville
0  01330 Villars les Dombes      
1  06364, NICE CEDEX 4           
2  10120                         
3  10400 Nogent sur seine        
4  113Lyon Saint Exupéry Aéroport
5  11890 Carcassonne cedex       
6  11890 Carcassonne cedex       
7  12101 Millau                  
8  18000 Bourges                 
9  21033 dijon cedex

**Ce qu'on y voit, c'est l'implémentation par les acheteurs de la colonne «ville» quand on classe *tout* le tableau par ordre alphabétique.**  
**Si vous lisez cet article sur JupyterLab (voir en bas de page pour plus d'infos), je vous invite à vous balader un peu plus loin dans la base, c'est parfois cocasse.**


___

#### **En attendant, par pure curiosité, voici première ville correctement entrée: c'est «ABLIS» en position ... 117.**

In [13]:
df_boamp_2018_classe_par_nom_de_ville.iloc[117,:]

d01_08_ville    ABLIS
Name: 117, dtype: object

___

# **B - Conséquences**

Tous les jeux de données sont à remanier, nettoyer, etc.  C'est comme ça.

Toujours est-il que ce nettoyage est plus ou moins compliqué à faire:
- Il faut unifier les casses;
- Faire attention aux caractères spéciaux;
- Se débarrasser des espaces superflus;
- ...

**Et après seulement**, on peut espérer travailler sur les données.

À titre d'exemple, si vous souhaitez géocoder vos annonces à partir de cette colonne, toutes les entrées avec un cedex ou toute autre décoration sera ignorée  
par les principaux outils à disposition et vos cartes seront incomplètes.  
Vous pourrez vous rabattre sur les codes postaux et les adresses, mais vous y rencontrerez sans doute des problèmes similaires.

____

# **C - Conclusion**

L'usage d'un bon référentiel d'implémentation de la base fait en effet défaut au BOAMP.

De ce fait, on l'a vu, de nombreuse données de marchés publics sont rendues (encore) plus absconses et deviennent rapidement inutilisables.  

**La cible de transparence voulue et inscrite dans les textes par le législateur, voire le constituant, est de ce fait manquée.**

#### **Mais gageons que tout cela va s'arranger, tant les projets autour de l'open data sont nombreux et les acteurs impliqués de plus en plus attentifs et volontaires (Etalab, Dila, ...).**

____

____

### **Bonus, aller plus loin avec ces données**

Pour m'éviter d'être assommant avec le code dans cet article, je n'ai pas poussé les manipulations du jeu de données.  
Si vous voulez vous amuser de votre côté avec ces 156 000 entrées sans avoir à télécharger tout un environnement  
**et si vous vous y entendez un petit peu sur JupyterLab et en Python**, voici la démarche à suivre:
- 1 - Rendez-vous sur Binder à cette adresse : https://mybinder.org/v2/gh/Semaine52/Semaine52Lab/master?urlpath=lab/tree/CONSOLE_PYTHON_A.ipynb  

NB / Le chargement peut prendre un moment, cela dépend de l'horaire.

- 2 - Copiez / Collez / Exécutez le code ci-dessous dans une cellule du Notebook qui va s'ouvrir directement:

In [14]:
# IMPORTS DES OUTILS
from urllib.request import urlretrieve
import pandas as pd

# RÉGLAGE IPYTHON INTERACTVESHELL
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all" 

# RÉGLAGES PANDAS
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_colwidth', -1)
pd.set_option('display.max_rows', 2000)
pd.set_option('display.max_columns', 500)

# IMPORT ET LECTURE DES DONNÉES
URL = 'http://www.semaine52.com/AuFilDuBoamp/NOTEBOOKS/S52_BILLET_2019_09_20_JMF_BOAMP_VILLE_CEDEX.ipynb'
urlretrieve(URL, 'BILLET_2019_09_20_JMF_BOAMP_BESOIN_REFERENTIEL.ipynb');

____

Attendez quelques secondes, et ce notebook apparaitra dans la liste des dossiers sur la gauche de votre écran, prêt à être ouvert. (cf. illustration ci-dessous)


![Importer et exécuter ce notebook sur Binder](https://www.semaine52.com/AuFilDuBoamp/IMG/illustration_import_binder.png)

À l'issue de vos expériences, vous pourrez naturellement faire un download de votre travail.

____

Merci de me faire part de toutes vos suggestions d'améliorations pour ce Notebook et/ou de me transmettre vos idées de sujets à traiter pour un prochain Billet.

Jean-Marie Falvet, Aix-en-Provence le 20 septembre 2019.  
Contact: https://twitter.com/semaine52

____